asapdiscovery-ml/asapdiscovery/ml/utils.py

```diff
@@ -1220,6 +1244,171 @@ def split_molecules(ds, split_fracs, generator=None):
     return all_subsets
 
 
+def split_temporal(ds, split_fracs, grouped=False, reverse=False, end_splits=2):
+    """
+    Split molecules temporally by date created. Earlier molecules will be placed in the
+    training set and later molecules will be placed in the val/test sets (unless
+    `reverse` is set).
+
+    Parameters
+    ----------
+    ds : Union[cml.data.DockedDataset, cml.data.GraphDataset]
+        Molecular dataset to split
+    split_fracs : List[float]
+        List of fraction of compounds to put in each split
+    grouped : bool, default=False
+        Splitting a GroupedDockedDataset object
+    reverse : bool, default=False
+        Reverse sorting of data
+    end_splits : int, default=2
+        How many splits (starting from the end of `split_fracs`) should be taken from
+        the end of the list (most recent data, or oldest if `reverse` is True). This is
+        in order to allow for the splits to not add up to 1, but still evaluate on the
+        most recent data, which can be used to assess how much temporal data is required
+        to predict on the most recent. Set to 0 to take all splits contiguously
+
+    Returns
+    -------
+    List[torch.utils.data.Subset]
+        List of Subsets of original dataset
+    """
+    import torch
+
+    # Get which splits should be from the front and which should be from the end
+    front_split_fracs = np.asarray(split_fracs[:-end_splits])
+    # Reverse the end splits so it matches when we reverse everything later
+    end_split_fracs = np.asarray(split_fracs[-end_splits:][::-1])
+
+    # Calculate how many molecules we want covered through each split
+    # By the end of each split, we should cumulatively covered the cumulative sum of all
+    #  splits up to and including that one
+    n_mols_front_split = front_split_fracs * len(ds)
+    # Same but for reverse
+    n_mols_end_split = end_split_fracs * len(ds)
+
+    # TODO: make this whole process more compact
```

Comment:
I think I have come up with a much more compact way of doing this, we will just need to work on what we want out of the `end_splits` arg, but this should be much more compact. 

Apologies for providing this in script format but  didn't want to push overwrite your hard work before we had a chance to discuss. 

```python

import pickle
import asapdiscovery
import numpy as np
import torch 

from asapdiscovery.ml.utils import split_temporal, split_molecules

def load_ds_from_pickle(path_to_pickle):
    with open(path_to_pickle, 'rb') as f:
        ds = pickle.load(f)
    return ds

ds = load_ds_from_pickle('blah.pkl')
key_func = lambda x: x["date_created"]
sorted_data_indices = sorted(range(len(ds)), key=lambda i: key_func(ds.structures[i]))

def split_list_into_chunks_range(data, ratios):
    chunks = []
    total_items = len(data)
    start = 0
    for ratio in ratios:
        end = int(start + ratio * total_items)
        chunks.append(data[start:end])
        ret_start = start
        start = end
    # tack any remainder onto the last chunk
    if end != total_items:
        chunks[-1] = data[ret_start: total_items]
    return chunks

chunks =  split_list_into_chunks_range(sorted_data_indices, [0.8, 0.1, 0.1])

ds_train = torch.utils.data.Subset(ds, chunks[0])
ds_test = torch.utils.data.Subset(ds, chunks[1])
ds_val = torch.utils.data.Subset(ds, chunks[2])

print("ds_train: ", ds_train, "length: ", len(ds_train))
print("ds_test: ", ds_test, "length: ", len(ds_test))
print("ds_val: ", ds_val, "length: ", len(ds_val))

# have a look at the dates
print("DS TRAIN\n")
for i in ds_train:
    print(i[1]["date_created"])

print("DS TEST\n")
# have a look at the dates
for i in ds_test:
    print(i[1]["date_created"])


print("DS VAL\n")
# have a look at the dates
for i in ds_val:
    print(i[1]["date_created"])
```

Reply:
this is definitely much more compact, but I think the main thing this is missing is the ability to have non-contiguous subsets, eg if we want to train on the earliest 30% of the molecules, but use the most recent 20% for the val/test splits. in this case, we would set `split_fracs=[0.3, 0.1, 0.1]` and `end_splits=2`, which means the last 2 entries in the `split_fracs` list should be taken from the end of the list of molecules sorted by date.

as an example, if we have our sorted list of molecules as `[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`, and we want the above split of training on the first 0.3 of molecules but test/val of the last 0.1/0.1, your code would give us splits of `[[1, 2, 3], [4], [5, 6, 7, 8, 9, 10]]` (the last split gets all leftover data). using my code, we'd get splits of `[[1, 2, 3], [9], [10]]`
(hopefully this makes things more clear and not more confusing lol)

Reply:
I think there's room to combine both of our implementations though, so let's definitely find a time to discuss!